### ISSUE TITLE
ClassNotFoundException when upgrading from 3.6.3 to 3.7.3

### ISSUE THREAD

**Issue by OskarKjellin** [hazelcast/hazelcast#9332]
We have a small cluster of 3 nodes running 3.6.3. Our upgrade process is to add new nodes running the new code and then stop the old ones. 
The new ones will never come to life to, making our rollback process trigger.

We're running java 8 on coreos/docker if that's of interest.

The existing nodes log:

```
[ippppppp]:5701 [xxxxxxxx] [3.6.3] Problem while reading DataSerializable, namespace: 0, id: 0, class: 'com.hazelcast.internal.cluster.impl.operations.MasterDiscoveryOperation', exception: com.hazelcast.internal.cluster.impl.operations.MasterDiscoveryOperation
com.hazelcast.nio.serialization.HazelcastSerializationException: Problem while reading DataSerializable, namespace: 0, id: 0, class: 'com.hazelcast.internal.cluster.impl.operations.MasterDiscoveryOperation', exception: com.hazelcast.internal.cluster.impl.operations.MasterDiscoveryOperation
	at com.hazelcast.internal.serialization.impl.DataSerializer.read(DataSerializer.java:130) ~[hazelcast-3.6.3.jar:3.6.3]
	at com.hazelcast.internal.serialization.impl.DataSerializer.read(DataSerializer.java:47) ~[hazelcast-3.6.3.jar:3.6.3]
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:46) ~[hazelcast-3.6.3.jar:3.6.3]
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:170) ~[hazelcast-3.6.3.jar:3.6.3]
	at com.hazelcast.spi.impl.NodeEngineImpl.toObject(NodeEngineImpl.java:234) ~[hazelcast-3.6.3.jar:3.6.3]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:378) ~[hazelcast-3.6.3.jar:3.6.3]
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.processPacket(OperationThread.java:184) ~[hazelcast-3.6.3.jar:3.6.3]
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.process(OperationThread.java:137) ~[hazelcast-3.6.3.jar:3.6.3]
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.doRun(OperationThread.java:124) ~[hazelcast-3.6.3.jar:3.6.3]
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.run(OperationThread.java:99) ~[hazelcast-3.6.3.jar:3.6.3]
Caused by: java.lang.ClassNotFoundException: com.hazelcast.internal.cluster.impl.operations.MasterDiscoveryOperation
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1333) ~[catalina.jar:8.0.37]
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1167) ~[catalina.jar:8.0.37]
	at com.hazelcast.nio.ClassLoaderUtil.tryLoadClass(ClassLoaderUtil.java:137) ~[hazelcast-3.6.3.jar:3.6.3]
	at com.hazelcast.nio.ClassLoaderUtil.loadClass(ClassLoaderUtil.java:115) ~[hazelcast-3.6.3.jar:3.6.3]
	at com.hazelcast.nio.ClassLoaderUtil.newInstance(ClassLoaderUtil.java:68) ~[hazelcast-3.6.3.jar:3.6.3]
	at com.hazelcast.internal.serialization.impl.DataSerializer.read(DataSerializer.java:119) ~[hazelcast-3.6.3.jar:3.6.3]
	... 9 more
```

And the new ones log:
```
[ipppppppp]:5701 [xxxxxxx] [3.7.3] No Invocation found for error response with callId: 0 sent from [ipppppppp]:5701
```

**Comment by jerrinot** [comment]
Hazelcast 3.6 and 3.7 have very different internals. Rolling Update is supported on a patch-level only.

**Comment by OskarKjellin** [comment]
Hmm, OK. Then I guess I was fooled by the release notes.
I understand that it might not be really applicable to what we use, but the property hazelcast.compatibility.3.6.client really sounded like it was supposed to provide compatibility towards 3.6. 
http://docs.hazelcast.org/docs/release-notes/
```
Hazelcast 3.6.x client cannot obtain cache proxy from 3.7.x cluster. Configuration creation and retrieval seem to be not working on 3.6.x clients when they connect to a 3.7.x cluster. Two new system properties added, namely hazelcast.compatibility.3.6.server and hazelcast.compatibility.3.6.client. When upgrading from (assuming that you are using ICache) 3.6.x to 3.7.3 then the servers should be started with -Dhazelcast.compatibility.3.6.client=true to enable existing 3.6.x clients work properly. When upgrading from 3.7.1 and 3.7.2 to 3.7.3 this is not needed.[9006]
```

**Comment by jerrinot** [comment]
@OskarKjellin: I am very sorry for the confusion. It's actually more complicated: There is a difference between client-server and server-server communication protocols. 

When you are using client/server topology then any 3.6+ Hazelcast client can talk to any 3.6+ server (3.6, 3.7, 3.8,...)  -> Hazelcast server-side can have a different lifecycle from your application (with Hazelcast clients)

However all servers in a same cluster have to be on the same minor version. They can differ in a patch level only. 

In Hazelcast 3.7 was a regression where 3.6 client was unable to obtain a JCache proxy from a newer server. This property `hazelcast.compatibility.3.6.clien` was introduced as a workaround for this regression.

**Comment by jerrinot** [comment]
I agree this should be documented better. cc @Serdaro

**Comment by Serdaro** [comment]
👍

**Comment by OskarKjellin** [comment]
Alright, thanks a lot for the support. I guess I'll stay on 3.6 then.

Thanks!

**Comment by jerrinot** [comment]
@OskarKjellin: you are very welcome. 

I'd suggest to move to 3.7.x whenever you can - there are tons of improvements -  I understand it's not always easy. Alternatively you could wait for 3.8 - Early Access should be available very soon.

**Comment by OskarKjellin** [comment]
Yeah I fully understand. There are some fixes in 3.7 that we really want, but not being able to do rolling upgrades kind of stop us from going there. I'll see if I can find a workaround
